src/main/kotlin/com/lovelysystems/gradle/S3DownloadFile.kt

```diff
@@ -0,0 +1,97 @@
+package com.lovelysystems.gradle
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Optional
+import org.gradle.api.tasks.TaskAction
+import software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.s3.S3AsyncClient
+import software.amazon.awssdk.services.s3.model.HeadBucketRequest
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request
+import software.amazon.awssdk.services.s3.model.NoSuchBucketException
+import software.amazon.awssdk.transfer.s3.S3TransferManager
+import software.amazon.awssdk.transfer.s3.model.DownloadFileRequest
+import java.io.File
+import java.nio.file.Files
+import java.nio.file.Paths
+import java.util.concurrent.ExecutionException
+import kotlin.io.path.createFile
+
+abstract class S3DownloadFile : DefaultTask() {
+
+    @get:Input
+    abstract val profile: Property<String>
+
+    @get:Input
+    abstract val bucket: Property<String>
+
+    @get:Input
+    abstract val targetFile: Property<File>
+
+    @get:Input
+    @get:Optional
+    abstract val region: Property<Region>
+
+    @get:Input
+    @get:Optional
+    abstract val key: Property<String>
+
+    init {
+        region.convention(Region.EU_CENTRAL_1)
+    }
+
+    @TaskAction
+    fun download() {
+        if (targetFile.get().exists()) {
+            logger.debug("Target file {} exists. Skipping download from S3.", targetFile.get())
```

Comment:
Added the right annotations so gradle can in most cases already give the right "up-to-date". However adding a simple upToDateWhen conditions isn't enough when e.g. the inputs of the task changed, for that case this custom abort takes effect. Alternativ would be to only rely on the gradle upToDateCheck and delete the file if it already exists. But that would cause more "unnecessary" downloads.

Reply:
good point. i think we don't know if the source changed with the config changes, i would say we should delete and download again. this log is not visible to the end user and so the config change would actually just be ignored.